actioncable/lib/action_cable.rb

```diff
@@ -41,15 +41,18 @@ module ActionCable
     disconnect_reasons: {
       unauthorized: "unauthorized",
       invalid_request: "invalid_request",
-      server_restart: "server_restart"
+      server_restart: "server_restart",
+      server_stop: "server_stop"
     },
     default_mount_path: "/cable",
     protocols: ["actioncable-v1-json", "actioncable-unsupported"].freeze
   }
 
   # Singleton instance of the server
   module_function def server
-    @server ||= ActionCable::Server::Base.new
+    @server ||= ActionCable::Server::Base.new.tap do |server|
+      at_exit { server.shutdown }
+    end
```

Comment:
Should server shutdown be triggered by the Rack handler? Or e.g. by a Puma `on_worker_shutdown` hook. Not sure whether there's any standard way to hook into Rack handler lifecycle. But an `at_exit` hook here is a good sign we aren't managing lifecycle at all 😅

Reply:
> by a Puma `on_worker_shutdown` hook

Sounds good. Thanks for the suggestion.

> an `at_exit` hook here is a good sign we aren't managing lifecycle at all

Yeah) That was just a proof of concept)
We were kinda surprised that there is no _canonical_ Rails way to implement graceful shutdown (e.g. `ActiveSupport.on_shutdown` hook?)

Reply:
@matthewd @tenderlove thoughts on who/what should take on lifecycle responsibility?

Interesting that Rack SPEC doesn't speak to this.

Reply:
@jeremy @palkan That's definitely the right point. The current idea is to remove `at_exit` and simply add some doc entries on how users can shutdown cable gracefully. For Puma, as you mentioned, they can use `on_worker_shutdown`, for Passenger there is a symmetrical `stopping_worker_process`  event.

As for the future, I guess we can add `ActiveSupport.on_shutdown` and instructions on how to integrate it with popular application servers. For Puma, we can even add the hook to bootstrapped config template.

Reply:
@jeremy @palkan Unfortunately, there is an issue with that. 😞 Puma, our major default application server, doesn't have a  reliable shutdown hook for all different configurations. `on_worker_shutdown` [works for cluster](https://github.com/puma/puma/blob/0b15d4ff6c94a6b144d6f61227672609f0f00175/lib/puma/dsl.rb#L336-L341) [mode only](https://github.com/puma/puma/blob/0991396844feea10e387bfaeabaacfd2dda26144/lib/puma/cluster.rb#L312). It means that, for instance, for default development configuration we can't rely on it, it's not invoked. 

Closing ActionCable from `ActiveSupport.on_shutdown` is definitely a good idea but we should invoke shutdown hooks somehow and I don't see a better option than notorious `at_exit` somewhere at the global level of Rails/ActiveSupport. WDYT?